Function support

jaredoconnell opened the pull request:
## Changes introduced with this PR

This PR adds functions to the schema.

There are two main categories of functions. Statically typed and dynamically typed.
- The way that the dynamically typed functions are implemented mean that we still validate their type, but a little looser using a function that determines the type after the input parameters are passed in. 
- Dynamically typed functions allow functions that support more generic things, like creating an object with specific fields.
- The statically typed functions are validated strictly. You pass in the function handler, and it's validated at runtime.
- The allowed return types for a standard function may or may not include an error return. This error return is in addition to the single return specified by the schema, if present.

For examples on how this is used, see the expressions draft PR: https://github.com/arcalot/arcaflow-expressions/pull/16

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

webbnh reviewed (commented):
Sorry that this is late -- I had a busy day -- and I've run out of time without getting to the tests...but I've got some comments, suggestions, and a question or two.

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+	return s.InputsValue
+}
+
+func (s CallableFunctionSchema) Output(inputType []Type) (Type, error) {
+	if s.DynamicTypeHandler == nil {
+		return s.DefaultOutputValue, nil
+	} else {
+		return s.DynamicTypeHandler(inputType)
+	}
+}
```
I agree with renaming it to Static. That's definitely the right way to go.

No, it should not use InputsValue. Because inputType is generally supposed to be more specific than InputsValue. That's what makes it dynamic. For example, if the input type specified in the schema is Any, then you can use the dynamic typing to detect int, str, map, etc.

webbnh commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+}
+
+func (f FunctionSchema) String() string {
+	result := f.ID() + "("
+	for i := 0; i < len(f.Parameters()); i++ {
+		if i != 0 {
+			result += ", "
+		}
+		result += string(f.Parameters()[i].TypeID())
+	}
+	result += ") "
```
> it would be more work to make that work

Yes it would, the _first_ time...but after that, you reap the rewards from the investment.

That is, if you had a `Map()` function, then I believe that you could code this like this:
```go
	guts := Map(f.Parameters(), func(p Type) string { return string(p.TypeID()) })
	result := f.ID() + "(" + ", ".join(guts) + ") "
```
which is probably better code (shorter, fewer moving parts, DRYer, better encapsulation, yadda yadda yadda).

The `Map()` function is easy to write (there are examples [here](https://stackoverflow.com/questions/33726731/short-way-to-apply-a-function-to-all-elements-in-a-list-in-golang), including a pointer to a package that you could import), and once it is available to your code, you might find other applications of it (even beyond `CallableFunctionSchema`).

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+func (s CallableFunctionSchema) Display() Display {
+	return s.DisplayValue
+}
+
+func (f CallableFunctionSchema) String() string {
```
I'll change them all to `f`.

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+	return result
+}
+
+func (s CallableFunctionSchema) ToFunctionSchema() (*FunctionSchema, error) {
+	if s.DynamicTypeHandler != nil && s.DefaultOutputValue == nil {
+		return nil, fmt.Errorf(
+			"function '%s' cannot be represented as a FunctionSchema because function has dynamic typing",
+			s.ID())
+	}
```
Yeah, I guess a fallback to static does not make sense here. I'll change it.

webbnh reviewed (commented):
More comments, mostly questions.  (There is one code comment which needs fixing, though.)

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,330 @@
+		handlerType := handler.Type().In(i)
+		if expectedType != handlerType {
+			return fmt.Errorf(
+				"type mismatch for parameter at index %d. handler has %v, inputs specifies %v",
```
I'm referencing the inputs schema slice. I'll try to clarify.

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,330 @@
+
+	switch {
+	case returnCount != 2:
+		return nil, fmt.Errorf("expected dynamic handler to have two returns, one any and one error, but got %d return types", returnCount)
```
I'll reword this.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+			passedInVal = test
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{"a"})
```
I don't think there's a need to overcomplicate the unit tests. There are no variables that would make it so that mechanisms tested in the prior test differ here.

webbnh replied:
I don't think there is _ever_ a need for **over**-complication!  🤣 

That aside....

> There are no variables that would make it so that mechanisms tested in the prior test differ here.

_You_ know that, and you know that it is true _at the moment_; however, the _test_ (e.g., "as a specification") doesn't "know" that, and it probably shouldn't presume it.  And, we hope that the test code will be longer-lived than the source code that it tests, such that known truths today many change in the future.  So, we really should be testing the behavior that we intend the code to follow, independent of any knowledge that we currently have of its actual implementation.

Also, test code has been known to evolve as well, and tests are usually evaluated on a per-test basis rather than as a part of a whole -- that is, it would be unfortunate if the previous test were removed or modified such that we lost the only place where that behavior was validated.  Another way of putting this is, each test should test "one thing", and it should be clear what its purpose is.  Thus, if the not-called-early behavior is a thing unto itself, it should have a separate test; otherwise, it should be checked by each of the tests to which it is pertinent.

jaredoconnell replied:
In the case that uses passedInVal, the function doesn't have a way to output its input, since it isolates just input in this test specifically. Once we check for that, that test is enough. I do not think it's worth it for the potential evolution.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+
+// Simple, with no input, and only error out.
+func TestCallableFunctionSchema_SimpleWithNilErr(t *testing.T) {
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() error {
+			return nil
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
```
They are identical. That's just what I ended up using. Sometimes the IDE complains when I use the second format. But I used that format for all of the calls to `Call`.

webbnh replied:
I really doubt that they are _identical_ -- knowing some of the Go philosophy, it seems highly unlikely that they would provide more than one way to do the same thing.  That said, I'm not finding any obvious difference between them in the language specification:  `make([]any, 0)` returns an initialized slice, backed by a "real", hidden, zero-length array, whereas `[]any{}` is a "slice literal", evaluating to an initialized, zero-length slice backed by a "real" (and also "hidden") array (which is presumably also zero-length).  I suspect that the difference is that (barring optimizations), the literal is allocated on the stack or perhaps in the process's static data region (there's lots of room there for zero-length things... 😉), whereas the result of `make()` is allocated from the heap.  (The Go documentation seems to be pretty non-specific about where and how they store things, but the fact that you can take the address of a local variable and it remains valid after leaving the function scope indicates that the implementation can and will allocate things from the heap when needed, even when they otherwise would be allocated on the stack in other languages.)

The fact that your IDE complains about certain usages is a further indication that they are _not_ identical.  However, I couldn't find any guidance on which to use when.  😬  It would be nice if we could follow some sort of consistency....

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+		make([]schema.Type, 0),
+		schema.NewStringSchema(nil, nil, nil),
+		nil,
+		func() string {
+			return "a"
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
```
This is generic, but also it could return an error if there is an incorrect number of args passed in, or if the value stored in the struct was not validated (by passing the checks in `NewCallableFunction` or `NewDynamicCallableFunction`).

`TestCallableFunctionSchema_Err_CallWrongErrorReturn` and `TestCallableFunctionSchema_Err_CallWrongReturnCount` test bypassing the checks.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+		},
+	)
+	assert.NoError(t, err)
+	_, err = simpleFunc.Call([]any{int64(1), int64(2)}) // Two args specified here
+	// There should be an error. Validate that it's the correct one.
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "incorrect number of args")
```
At the moment, not programmatically.
Theoretically, I could create an error type that reports that information if you think it would be valuable.

webbnh replied:
As we discussed over DM, this seems like an interface "smell".

What _is_ the purpose of the error value returned by the user function, who is supposed to consume that value, and what behavior is it supposed to trigger?

If the expectation is that the engine/SDK is the consumer, then that largely decouples the error returned by `Call()` from the error returned by the user function.  That is, if the user function returns an error, then presumably `Call()` will return one as well, but that is just one of a number of outcomes which would prompt `Call()` to return an error.  (And, the details of how this outcome should be conveyed back to the calling user code should be explored, including a means of making it clear to its caller why `Call()` is returning an error under these circumstances.)  

However, if the calling user code has some expectation of being able to see the error value returned by the user function, then we really should not be using the same output parameter to report errors from both `Call()` and the user function.

Does that make sense?

jaredoconnell replied:
The newest commit adds a way to differentiate the error.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+	_, err := callable.Call(make([]any, 0))
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "error return val isn't an error")
+}
```
In Golang, the protections for this are weak. Anyone can instantiate the struct directly, bypassing the protections. So I am just adding coverage for that case. It's not recommended, though.

webbnh replied:
I think you need to draw a line and say, "this is what we support" (and that doesn't include instantiating the structures except through the published interfaces).  There needs to be some set of acceptable practices within which you try to test everything; this looks like it falls outside of acceptable practice...if you are going to test _this_ case, what other cases should you also be testing?

I'm pretty sure that in this way lies madness....

jaredoconnell commented on schema/function_test.go:
```diff
@@ -515,3 +515,68 @@ func TestNewDynamicFunction_SliceOfSameType(t *testing.T) {
+	assert.Equals(t, funcStr, "a(string) void")
+}
+
+func TestFunctionToStringTwoParam(t *testing.T) {
+	oneParamVoidFunction, err := schema.NewCallableFunction(
+		"b",
+		[]schema.Type{
+			schema.NewStringSchema(nil, nil, nil),
+			schema.NewIntSchema(nil, nil, nil),
+		},
+		schema.NewIntSchema(nil, nil, nil),
+		nil,
+		func(a string, b int64) int64 { return 0 },
+	)
+	assert.NoError(t, err)
+	funcStr := oneParamVoidFunction.String()
+	assert.Equals(t, funcStr, "b(string, integer) integer")
+}
+
+func TestToFunctionSchemaAndString(t *testing.T) {
+	oneParamVoidCallableFunction, err := schema.NewCallableFunction(
+		"b",
+		[]schema.Type{
+			schema.NewStringSchema(nil, nil, nil),
+			schema.NewIntSchema(nil, nil, nil),
+		},
+		schema.NewIntSchema(nil, nil, nil),
+		nil,
+		func(a string, b int64) int64 { return 0 },
+	)
+	assert.NoError(t, err)
+	oneParamVoidFunction, err := oneParamVoidCallableFunction.ToFunctionSchema()
+	assert.NoError(t, err)
+
+	funcStr := oneParamVoidFunction.String()
+	assert.Equals(t, funcStr, "b(string, integer) integer")
+}
```
I'll combine them, and add a validation for failure to convert to FunctionSchema in the dynamic function test that follows.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -515,3 +515,68 @@ func TestNewDynamicFunction_SliceOfSameType(t *testing.T) {
+		"c",
+		[]schema.Type{schema.NewAnySchema(), schema.NewIntSchema(nil, nil, nil)},
+		nil,
+		func(a any, b int64) (any, error) {
+			return a, nil
+		},
+		func(inputType []schema.Type) (schema.Type, error) {
+			return schema.NewIntSchema(nil, nil, nil), nil
+		},
```
I'll change it to return `b`.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+			return result.Interface(), nil
+		},
+		func(inputTypes []schema.Type) (schema.Type, error) {
+			if len(inputTypes) != 1 {
+				return nil, fmt.Errorf("expected 1 arg, got %d", len(inputTypes))
+			}
```
I don't think so. What's there will just help in case someone changes the test incorrectly.

webbnh replied:
In that case, shouldn't the result be a little "firmer"?  Perhaps instead of returning, it should fail an assertion or otherwise crash the test?  (Depending on the test itself to notice the error returned from here is somewhat weak.)

jaredoconnell replied:
Sure. Switched to assertion.

webbnh commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
+	// No result or error
+	assert.NoError(t, err)
+	assert.Nil(t, result)
```
So, I think the answer is that the `Call()` interface returns two values, a result and an error value:  the result will be the value returned by the called function if it returns one; if the function is not successfully invoked, or if the function returned no value, then the result will be `nil`.  Since, in this test scenario, the function does not return a value, the value of `result` should be `nil`, and it is appropriate for the test code to check that.

webbnh reviewed (commented):
Other than some questions around handling functions which return errors, this looks like it's converging on completion.

I do have a couple of nits for you on your update, and I don't know if you saw [this question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1420841261).

By the way, I don't seem to have the power to "resolve" conversations which are no longer an issue...how should we go about closing them?

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,591 @@
+}
+
+func TestFunctionToStringToFunctionTwoParam(t *testing.T) {
+	oneParamVoidCallableFunction, err := schema.NewCallableFunction(
+		"b",
+		[]schema.Type{
+			schema.NewStringSchema(nil, nil, nil),
+			schema.NewIntSchema(nil, nil, nil),
+		},
+		schema.NewIntSchema(nil, nil, nil),
+		nil,
+		func(a string, b int64) int64 { return 0 },
+	)
```
Good catch. Probably a copy/paste issue. Will be fixed on the next commit.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+			passedInVal = test
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{"a"})
```
Go wouldn't allow that. It's fairly strict about typing. A slice of string cannot be passed into a slice of any.

webbnh reviewed (approved):
Looks good, although I found one more nit on this last pass.

So, the lingering issues are these:

- [x] There is the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1420979615) of whether the function schema should describe explicitly whether the function should be expected to return an error value in addition to the specified output, if any; and there's the related [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423026178) of whether that error value is expected to be consumed by the SDK/engine or relayed to the user code and the details of how.
- [x] There's the [suggestion](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1419609439) to rework the function validation function(s).
- [x] There's the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1422852792) of how much knowledge should be assumed by the test.
- [x] There's the [musing](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423004711) over when to use slice literals vs. `make()` which I'm hoping you can illuminate.
- [x] There's the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423031846) of what you actually need to test (and whether it is reasonable, independent of what's possible, for the caller to "go around" the interface and use objects created directly).
- [ ] There's my [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423035739) about how the code in the test which tests the test should be reporting failure.
- [ ] And, there's a general suggestion about relying more on function header comments and less on line comments (particularly when the line comments don't add much beyond what the code already says).

None of these rise to the level of blockers (so I've approved), but I would appreciate at least a comment from you on each of them.

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+	if err != nil {
+		return nil, err
+	}
+	// Validate the output type
+	if output == nil {
+		err := validateVoidFunc(parsedHandler)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err := validateTypedReturnFunc(parsedHandler, output)
+		if err != nil {
+			return nil, err
+		}
+	}
```
I will push my changes soon. Let me know if you think the new function is designed well. It's designed similarly to Webb's proposal.

jaredoconnell commented on schema/function.go:
```diff
@@ -0,0 +1,326 @@
+	}, nil
+}
+
+func validateVoidFunc(parsedHandler reflect.Value) error {
+	returnCount := parsedHandler.Type().NumOut()
+	// A void function can have no returns or an error return
+	if returnCount > 1 {
+		return fmt.Errorf(
+			"parameter output is nil, meaning it's a void function, or a function with just an error return, but got %d return types",
+			returnCount,
+		)
+	} else if returnCount == 1 {
+		// Validate that it's just an error return
+		returnTypeName := parsedHandler.Type().Out(0).Name()
+		if returnTypeName != errorType {
+			return fmt.Errorf("expected void or error return, but got %s", returnTypeName)
+		}
+	}
+	return nil
+}
+
+func validateTypedReturnFunc(parsedHandler reflect.Value, outputType Type) error {
+	returnCount := parsedHandler.Type().NumOut()
+
+	if returnCount > 2 || returnCount < 1 {
+		return fmt.Errorf("expected handler to have one return, or one plus an error return, but got %d return types", returnCount)
+	} else {
+		// Validate the return type
+		expectedType := outputType.ReflectedType()
+		handlerType := parsedHandler.Type().Out(0)
+		if expectedType != handlerType {
+			return fmt.Errorf("mismatched return type. expected %s, handler has %s", expectedType, handlerType)
+		}
+		// Validate error return, if applicable.
+		if returnCount == 2 && parsedHandler.Type().Out(1).Name() != errorType {
+			return fmt.Errorf("expected additional return type to be an error return, but got %s", parsedHandler.Type().Out(1).Name())
+		}
+	}
+	return nil
+}
```
I will push my changes soon. For now I made it so that the Callable function knows whether or not there may be an error. Function schema doesn't.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+func TestNewCallableFunction_Err_MismatchedParamType(t *testing.T) {
+	_, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)}, // String specified here
+		nil,
+		nil,
+		func(int) {}, // Int specified here, mismatched
```
I think they make it easy for those who aren't experienced with the code base. I'll leave them in because the clutter they add isn't worse than the value they add, in Matt and I's opinions.

jaredoconnell commented on schema/function_test.go:
```diff
@@ -0,0 +1,517 @@
+			return a, nil
+		},
+		func(inputTypes []schema.Type) (schema.Type, error) {
+			if len(inputTypes) != 1 {
+				return nil, fmt.Errorf("expected 1 arg, got %d", len(inputTypes))
+			}
```
I don't think so. It's just in case someone breaks the test case.

jaredoconnell replied:
Switched to assertion based on your other comment.

jaredoconnell commented:
- There is the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1420979615) of whether the function schema should describe explicitly whether the function should be expected to return an error value in addition to the specified output, if any; and there's the related [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423026178) of whether that error value is expected to be consumed by the SDK/engine or relayed to the user code and the details of how.
  - Changed to make it explicit for CallableFunctionSchema only.
- There's the [suggestion](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1419609439) to rework the function validation function(s).
  - I did something similar to your proposal
- There's the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1422852792) of how much knowledge should be assumed by the test.
  - Responded. It's only to test the simple case where the function doesn't have a return value. It's not required for this test or any test.
- There's the [musing](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423004711) over when to use slice literals vs. `make()` which I'm hoping you can illuminate.
  - The behavior is identical. The allocations are similar enough in that they'll both end up on the heap due to the allocation going out of scope. It seems more like a convention thing, and as-written makes the linters and IDE happy.
- There's the [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423031846) of what you actually need to test (and whether it is reasonable, independent of what's possible, for the caller to "go around" the interface and use objects created directly).
  - As designed, nothing prevents someone from initializing their own struct. There are late-stage validations in there that should get coverage, so there are basic tests for that.
- There's my [question](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/68#discussion_r1423035739) about how the code in the test which tests the test should be reporting failure.
  - I switched to an assertion.
- And, there's a general suggestion about relying more on function header comments and less on line comments (particularly when the line comments don't add much beyond what the code already says).
  - I think it adds more to an unfamiliar person than it subtracts due to clutter.

webbnh commented on schema/function.go:
```diff
@@ -198,23 +216,36 @@ func (f FunctionSchema) Display() Display {
 }
 
+func getReturnTypeString(returnType Type, hasError bool) string {
+	switch {
+	case returnType != nil:
+		if hasError {
+			return "(" + string(returnType.TypeID()) + ", error)"
+		} else {
+			return string(returnType.TypeID())
+		}
+	case hasError:
+		return "error"
+	default:
+		return "void"
+	}
```
Again, only if you are making changes for some other reason...this might be more graceful code:
```golang
	switch {
	case returnType != nil && hasError:
		return "(" + string(returnType.TypeID()) + ", error)"
	case returnType != nil:
		return string(returnType.TypeID())
	case hasError:
		return "error"
	default:
		return "void"
	}
```

jaredoconnell replied:
I'll make the change.

webbnh commented on schema/function.go:
```diff
@@ -251,13 +283,10 @@ func (f CallableFunctionSchema) Display() Display {
-		result += string(f.StaticOutputValue.TypeID())
-	default:
-		result += "void"
+	if f.DynamicTypeHandler != nil {
+		result += "(dynamic, error)"
```
Shouldn't the presence of `error` here be conditional on the value of `f.OutputsError`?

jaredoconnell replied:
As coded, it expects an error return in addition to the type returned by the handler. So it doesn't use the OutputsError in that case.

jaredoconnell replied:
I improved the documentation.

webbnh commented on schema/function_test.go:
```diff
@@ -0,0 +1,664 @@
+	assert.Equals(t, funcStr, "b(string, integer) integer")
+}
+
+func TestDynamicFunctionToStringToFunction(t *testing.T) {
+	oneParamVoidFunction, err := schema.NewDynamicCallableFunction(
+		"c",
+		[]schema.Type{schema.NewAnySchema(), schema.NewIntSchema(nil, nil, nil)},
+		nil,
+		func(a any, b int64) (any, error) {
+			return b, nil
+		},
+		func(inputType []schema.Type) (schema.Type, error) {
+			return schema.NewIntSchema(nil, nil, nil), nil
+		},
+	)
+	assert.NoError(t, err)
+	funcStr := oneParamVoidFunction.String()
+	assert.Equals(t, funcStr, "c(any, integer) (dynamic, error)")
```
Do dynamic functions always/have-to return an error value?  (If not, should you have a test of one that doesn't?)

jaredoconnell replied:
They should. It's enforced by NewDynamicCallableFunction.

webbnh reviewed (approved):
Looks good.  I just have a couple of questions and a couple of suggestions (which you can consider or ignore if don't have any other changes to make).
